Source/SwiftLintFramework/Rules/Lint/RequiredXCTestTearDownRule.swift

```diff
@@ -0,0 +1,141 @@
+            final class FooTests: XCTestCase {
+                override func setUp() {}
+                override func tearDownWithError() throws {}
+            }
+            """#),
+            Example(#"""
+            final class FooTests: XCTestCase {
+                override func setUpWithError() throws {}
+                override func tearDownWithError() throws {}
+            }
+            final class BarTests: XCTestCase {
+                override func setUpWithError() throws {}
+                override func tearDownWithError() throws {}
+            }
+            """#),
+            Example(#"""
+            struct FooTests {
+                override func setUp() {}
+            }
+            class BarTests {
+                override func setUpWithError() throws {}
+            }
+            """#),
+            Example(#"""
+            final class FooTests: XCTestCase {
+                override func setUpAlLExamples() {}
+            }
+            """#),
+            Example(#"""
+            final class FooTests: XCTestCase {
+                class func setUp() {}
+                class func tearDown() {}
+            }
+            """#)
+        ],
+        triggeringExamples: [
+            Example(#"""
+            final class ↓FooTests: XCTestCase {
+                override func setUp() {}
+            }
+            """#),
+            Example(#"""
+            final class ↓FooTests: XCTestCase {
+                override func setUpWithError() throws {}
+            }
+            """#),
+            Example(#"""
+            final class FooTests: XCTestCase {
+                override func setUp() {}
+                override func tearDownWithError() throws {}
+            }
+            final class ↓BarTests: XCTestCase {
+                override func setUpWithError() throws {}
+            }
+            """#),
+            Example(#"""
+            final class ↓FooTests: XCTestCase {
+                class func setUp() {}
+            }
+            """#)
+        ]
+    )
+
+    // MARK: - Life cycle
+
+    public init() {}
+
+    // MARK: - Public
+
+    public func validate(file: SwiftLintFile) -> [StyleViolation] {
+        testClasses(in: file).compactMap { violations(in: file, for: $0) }
+    }
+
+    // MARK: - Private
+
+    private func testClasses(in file: SwiftLintFile) -> [SourceKittenDictionary] {
+        file.structureDictionary.substructure.filter { dictionary in
+            guard dictionary.declarationKind == .class else { return false }
+            return dictionary.inheritedTypes.contains("XCTestCase")
+        }
+    }
+
+    private func violations(in file: SwiftLintFile,
+                            for dictionary: SourceKittenDictionary) -> StyleViolation? {
+        let methods = dictionary.substructure
+            .compactMap { XCTMethod($0.name) }
+
+        guard
+            methods.contains(.setUp) == true,
+            methods.contains(.tearDown) == false,
```

Comment:
Perhaps this rule could be called `BalancedXCTestLifecycleRule` and also enforce that if `tearDown()` is defined, then so should `setUp()`?

```suggestion
            methods.contains(.setUp) != methods.contains(.tearDown),
```

Reply:
Exactly! I was going to say I will rename the rule, but just saw your comment :-) I'll rename it. Thanks!

Reply:
All changes applied, @jpsim. I'm merging it.